src/dotnet/BuildServer/BuildServerProvider.cs

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.DotNet.Configurer;
+using Microsoft.Extensions.EnvironmentAbstractions;
+
+namespace Microsoft.DotNet.BuildServer
+{
+    internal class BuildServerProvider : IBuildServerProvider
+    {
+        private readonly IFileSystem _fileSystem;
+        private readonly IEnvironmentProvider _environmentProvider;
+
+        public BuildServerProvider(
+            IFileSystem fileSystem = null,
+            IEnvironmentProvider environmentProvider = null)
+        {
+            _fileSystem = fileSystem ?? FileSystemWrapper.Default;
+            _environmentProvider = environmentProvider ?? new EnvironmentProvider();
+        }
+
+        public IEnumerable<IBuildServer> EnumerateBuildServers(ServerEnumerationFlags flags = ServerEnumerationFlags.All)
+        {
+            if ((flags & ServerEnumerationFlags.MSBuild) == ServerEnumerationFlags.MSBuild)
+            {
+                // Yield a single MSBuild server (handles server discovery itself)
+                // TODO: use pid file enumeration when supported by the server
+                yield return new MSBuildServer();
+            }
+
+            if ((flags & ServerEnumerationFlags.VBCSCompiler) == ServerEnumerationFlags.VBCSCompiler)
+            {
+                // Yield a single VB/C# compiler (handles server discovery itself)
+                // TODO: use pid file enumeration when supported by the server
+                yield return new VBCSCompilerServer();
+            }
+
+            // TODO: remove this check when Razor isn't the only server writing a pid file
+            if ((flags & ServerEnumerationFlags.Razor) != ServerEnumerationFlags.Razor)
+            {
+                yield break;
```

Comment:
One of the questions that came up on the original issue was whether the shutdown would be bound to commands associated with the current sdk version. e.g. would shutdown using SDK 2.2 shutdown components of SDK 2.1, etc. etc. And the answer, before considering razor, out-of-band compiler was that we would stick to the current SDK version. I think @dasMulli raised the question.

The PID process actually changes that, and I actually like this new semantic better.

But thought it might be worth having some public discussion on it...

It seems to me that this is what I always want when I say `dotnet build-server shutdown`: shut down *anything* that was created by a build that may be locking files, etc. I don't want to have to do this multiple times if I happen to have used more than 1 SDK SxS recently.

Furthermore, since, the shutdown request is graceful, I don't think the cross-talk between SDK versions is a real issue.

Reply:
I also would expect `dotnet build-server shutdown` to shut down any servers running, regardless of where or how they were started.

This is one of the reasons the pid file now has the path to the executable to use to shut down the server (rather than just a pipe name for dotnet to communicate with), so that it can use the same protocol that the running server is expecting.  Also keeps the protocol implementation out of `dotnet`.

I've filed bugs to hopefully move the Roslyn and MSBuild implementations in this direction for the future.